File: src/main/java/no/novari/flyt/history/model/event/EventCategorizationService.java
package no.novari.flyt.history.model.event;

import lombok.Getter;
import no.novari.flyt.history.model.instance.InstanceStatus;
import no.novari.flyt.history.model.instance.InstanceStorageStatus;
import no.novari.flyt.history.repository.filters.EventNamesPerInstanceStatus;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

@Getter
@Service
public class EventCategorizationService {

    private final EventNamesPerInstanceStatus eventNamesPerInstanceStatus;

    public EventCategorizationService() {
        eventNamesPerInstanceStatus = EventNamesPerInstanceStatus
                .builder()
                .inProgressStatusEventNames(getEventNamesByInstanceStatuses(InstanceStatus.IN_PROGRESS))
                .transferredStatusEventNames(getEventNamesByInstanceStatuses(InstanceStatus.TRANSFERRED))
                .abortedStatusEventNames(getEventNamesByInstanceStatuses(InstanceStatus.ABORTED))
                .failedStatusEventNames(getEventNamesByInstanceStatuses(InstanceStatus.FAILED))
                .allStatusEventNames(getAllInstanceStatusEventNames())
                .build();
    }

    public Set<EventCategory> getCategoriesByInstanceStorageStatuses(Collection<InstanceStorageStatus> instanceStorageStatuses) {
        return Arrays.stream(EventCategory.values())
                .filter(eventCategory -> instanceStorageStatuses.contains(eventCategory.getInstanceStorageStatus()))
                .collect(Collectors.toSet());
    }

    public Set<String> getEventNamesByInstanceStorageStatuses(Collection<InstanceStorageStatus> instanceStorageStatuses) {
        return getCategoriesByInstanceStorageStatuses(instanceStorageStatuses)
                .stream()
                .map(EventCategory::getEventName)
                .collect(Collectors.toSet());
    }

    public Set<EventCategory> getCategoriesByInstanceStatuses(Collection<InstanceStatus> instanceStatuses) {
        return Arrays.stream(EventCategory.values())
                .filter(eventCategory -> instanceStatuses.contains(eventCategory.getInstanceStatus()))
                .collect(Collectors.toSet());
    }

    public Set<String> getEventNamesByInstanceStatuses(InstanceStatus... instanceStatuses) {
        return getEventNamesByInstanceStatuses(Arrays.asList(instanceStatuses));
    }

    public Set<String> getEventNamesByInstanceStatuses(Collection<InstanceStatus> instanceStatuses) {
        return getCategoriesByInstanceStatuses(instanceStatuses)
                .stream()
                .map(EventCategory::getEventName)
                .collect(Collectors.toSet());
    }

    private final Map<String, EventCategory> categoryByName = Arrays.stream(EventCategory.values())
            .collect(Collectors.toMap(EventCategory::getEventName, Function.identity()));

    public EventCategory getCategoryByEventName(String name) {
        return categoryByName.get(name);
    }

    public InstanceStatus getStatusByEventName(String name) {
        EventCategory category = getCategoryByEventName(name);
        if (category == null) {
            throw new IllegalArgumentException("No category with name=" + name);
        }
        InstanceStatus status = category.getInstanceStatus();
        if (status == null) {
            throw new IllegalArgumentException("Category=" + name + " is not an instance status category");
        }
        return status;
    }

    public InstanceStorageStatus getStorageStatusByEventName(String name) {
        EventCategory category = getCategoryByEventName(name);
        if (category == null) {
            throw new IllegalArgumentException("No category with name=" + name);
        }
        InstanceStorageStatus status = category.getInstanceStorageStatus();
        if (status == null) {
            throw new IllegalArgumentException("Category=" + name + " is not an instance storage status category");
        }
        return status;
    }

    public Set<EventCategory> getInstanceStatusCategories() {
        return Arrays.stream(EventCategory.values())
                .filter(eventCategory -> Objects.nonNull(eventCategory.getInstanceStatus()))
                .collect(Collectors.toSet());
    }

    public Set<String> getAllInstanceStatusEventNames() {
        return getInstanceStatusCategories()
                .stream()
                .map(EventCategory::getEventName)
                .collect(Collectors.toSet());
    }

    public Set<EventCategory> getInstanceStorageStatusCategories() {
        return Arrays.stream(EventCategory.values())
                .filter(eventCategory -> Objects.nonNull(eventCategory.getInstanceStorageStatus()))
                .collect(Collectors.toSet());
    }

    public Set<String> getAllInstanceStorageStatusEventNames() {
        return getInstanceStorageStatusCategories()
                .stream()
                .map(EventCategory::getEventName)
                .collect(Collectors.toSet());
    }

}

File: src/main/kotlin/no/novari/flyt/history/model/event/EventCategorizationService.kt
package no.novari.flyt.history.model.event

import no.novari.flyt.history.model.instance.InstanceStatus
import no.novari.flyt.history.model.instance.InstanceStorageStatus
import no.novari.flyt.history.repository.filters.EventNamesPerInstanceStatus
import org.springframework.stereotype.Service

@Service
class EventCategorizationService {
    val eventNamesPerInstanceStatus: EventNamesPerInstanceStatus =
        EventNamesPerInstanceStatus
            .builder()
            .inProgressStatusEventNames(getEventNamesByInstanceStatuses(InstanceStatus.IN_PROGRESS))
            .transferredStatusEventNames(getEventNamesByInstanceStatuses(InstanceStatus.TRANSFERRED))
            .abortedStatusEventNames(getEventNamesByInstanceStatuses(InstanceStatus.ABORTED))
            .failedStatusEventNames(getEventNamesByInstanceStatuses(InstanceStatus.FAILED))
            .allStatusEventNames(allInstanceStatusEventNames)
            .build()

    private val categoryByName = EventCategory.entries.associateBy(EventCategory::eventName)

    fun getCategoriesByInstanceStorageStatuses(
        instanceStorageStatuses: Collection<InstanceStorageStatus>,
    ): Set<EventCategory> {
        return EventCategory.entries
            .filter { it.instanceStorageStatus in instanceStorageStatuses }
            .toSet()
    }

    fun getEventNamesByInstanceStorageStatuses(
        instanceStorageStatuses: Collection<InstanceStorageStatus>,
    ): Set<String> {
        return getCategoriesByInstanceStorageStatuses(instanceStorageStatuses)
            .mapTo(linkedSetOf(), EventCategory::eventName)
    }

    fun getCategoriesByInstanceStatuses(instanceStatuses: Collection<InstanceStatus>): Set<EventCategory> {
        return EventCategory.entries
            .filter { it.instanceStatus in instanceStatuses }
            .toSet()
    }

    fun getEventNamesByInstanceStatuses(vararg instanceStatuses: InstanceStatus): Set<String> {
        return getEventNamesByInstanceStatuses(instanceStatuses.asList())
    }

    fun getEventNamesByInstanceStatuses(instanceStatuses: Collection<InstanceStatus>): Set<String> {
        return getCategoriesByInstanceStatuses(instanceStatuses)
            .mapTo(linkedSetOf(), EventCategory::eventName)
    }

    fun getCategoryByEventName(name: String): EventCategory? = categoryByName[name]

    fun getStatusByEventName(name: String): InstanceStatus {
        val category = requireNotNull(getCategoryByEventName(name)) { "No category with name=$name" }
        return requireNotNull(category.instanceStatus) { "Category=$name is not an instance status category" }
    }

    fun getStorageStatusByEventName(name: String): InstanceStorageStatus {
        val category = requireNotNull(getCategoryByEventName(name)) { "No category with name=$name" }
        return requireNotNull(
            category.instanceStorageStatus,
        ) { "Category=$name is not an instance storage status category" }
    }

    val instanceStatusCategories: Set<EventCategory>
        get() = EventCategory.entries.filter { it.instanceStatus != null }.toSet()

    val allInstanceStatusEventNames: Set<String>
        get() = instanceStatusCategories.mapTo(linkedSetOf(), EventCategory::eventName)

    val instanceStorageStatusCategories: Set<EventCategory>
        get() = EventCategory.entries.filter { it.instanceStorageStatus != null }.toSet()

    val allInstanceStorageStatusEventNames: Set<String>
        get() = instanceStorageStatusCategories.mapTo(linkedSetOf(), EventCategory::eventName)
}
